FundTrack.BLL/Concrete/ReportService.cs

```diff
@@ -101,5 +102,96 @@ public IEnumerable<String> GetImagesById(int finOpId)
+            return t.ToList().Count;
+        }
+
+        public async Task<int> GetFilteredCountOfUsersDonationsReport(int orgId, DateTime dateFrom, DateTime dateTo, string filterValue)
+        {
+            var t = await GetUserDonations(orgId, dateFrom, dateTo);
+            return t.Where(d => d.UserLogin.Contains(filterValue)).ToList().Count;
+        }
+
+        public async Task<int> GetCountOfCommonUsersDonationsReport(int orgId, DateTime dateFrom, DateTime dateTo)
+        {
+            var t = await GetUserDonations(orgId, dateFrom, dateTo);
+            var res = GroupDonationsByUserLogin(t.ToList());
+            return res.ToList().Count;
+        }
+
+
+        public async Task<IEnumerable<UsersDonationsReportViewModel>> GetUsersDonationsPaginatedReport(int orgId, DateTime dateFrom, DateTime dateTo, int pageIndex, int pageSize)
+        {
+            var t = await GetUserDonations(orgId, dateFrom, dateTo);
+            return PaginateReport(pageIndex, pageSize, t);
+        }
+
+        public async Task<IEnumerable<UsersDonationsReportViewModel>> GetFilteredUsersDonationsPaginatedReport(int orgId, DateTime dateFrom, DateTime dateTo, int pageIndex, int pageSize, string filterValue)
+        {
+            var t = await GetUserDonations(orgId, dateFrom, dateTo);
+            var res = t.Where(d => d.UserLogin.Contains(filterValue));
+            return PaginateReport(pageIndex, pageSize, res);
+        }
+
+        public async Task<IEnumerable<UsersDonationsReportViewModel>> GetCommonUsersDonationsPaginatedReport(int orgId, DateTime dateFrom, DateTime dateTo, int pageIndex, int pageSize)
+        {
+            var t = await GetUserDonations(orgId, dateFrom, dateTo);
+            var res = GroupDonationsByUserLogin(t.ToList());
+            return PaginateReport(pageIndex, pageSize, res);
+        }
+
+        private static IQueryable<UsersDonationsReportViewModel> GroupDonationsByUserLogin(IEnumerable<UsersDonationsReportViewModel> list)
+        {
+            return list.GroupBy(d => d.UserLogin)
+                .Select(d => new UsersDonationsReportViewModel()
+            {
+                UserLogin = d.Select(u => u.UserLogin).FirstOrDefault(),
+                UserFulName = d.Select(u => u.UserFulName).FirstOrDefault() ,
+                MoneyAmount = d.Sum(u => u.MoneyAmount),
+                Target = String.Join(", ", d.Select(u=>u.Target).Where(t=> !String.IsNullOrEmpty(t)).Distinct().ToArray())
+            }).AsQueryable();
+        }
+
+        private List<UsersDonationsReportViewModel> PaginateReport(int pageIndex, int pageSize, IQueryable<UsersDonationsReportViewModel> t)
+        {
+            var list = t.Skip((pageIndex - 1) * pageSize).Take(pageSize).ToList();
+            list.Sort();
+            return list;
+        }
+
+        private Task<IQueryable<UsersDonationsReportViewModel>> GetUserDonations(int orgId, DateTime dateFrom, DateTime dateTo)
+        {
+            return Task.Run(() =>

```

Comment:
task.run executes action on another thread. but you actually do not need another thread here.